download-basket/src/main/java/downloadbasket/helpers/SendDownloadDetailsToEmailThread.java

```diff
@@ -0,0 +1,289 @@
+				JSONObject download = downLoadDetails.getJSONObject(i);
+				final String croppingMode = download.getString(PARAM_CROPPING_MODE);
+				String croppingLayer = "";
+				if (download.has(PARAM_CROPPING_LAYER)) {
+					croppingLayer = download.getString(PARAM_CROPPING_LAYER);
+				}
+
+				LoadZipDetails ldz = new LoadZipDetails();
+				ldz.setTemporaryDirectory(strTempDir);
+				ldz.setUserEmail(userDetails.getString("email"));
+				ldz.setLanguage(this.locale.getLanguage());
+				ldz.setDownloadNormalWay(normalDownloads.isBboxCropping(croppingMode, croppingLayer));
+				OskariLayerService mapLayerService = new OskariLayerServiceIbatisImpl();
+				OskariLayer oskariLayer = mapLayerService.find(download.getString(PARAM_LAYER_ID));
+				String srs = "EPSG:4326";
+				if (oskariLayer != null) {
+					srs = oskariLayer.getSrs_name();
+				}
+
+				if (ldz.isDownloadNormalWay()) {
+					ldz.setGetFeatureInfoRequest(OGCServices.getFilter(download, true));
+					ldz.setWFSUrl(OGCServices.doGetFeatureUrl(srs, download, false));
+				} else {
+
+					ldz.setGetFeatureInfoRequest("&filter=" +OGCServices.getPluginFilter(download));
+					ldz.setWFSUrl(OGCServices.doGetFeatureUrl(srs, download, true));
+				}
+
+				final String fileLocation = ds.loadZip(ldz, this.locale);
+
+				if (fileLocation != null) {
+					ZipDownloadDetails zdd = new ZipDownloadDetails();
+					zdd.setFileName(fileLocation);
+					final String sLayer = Helpers.getLayerNameWithoutNameSpace(download.getString(PARAM_LAYER));
+					zdd.setLayerName(sLayer);
+					mergeThese.add(zdd);
+				}
+			}
+
+			ZipOutputStream out = null;
+			String strZipFileName = UUID.randomUUID().toString() + ".zip";
+			try {
+				File f = new File(strTempDir);
+				f.mkdirs();
+				out = new ZipOutputStream(new FileOutputStream(new File(strTempDir, strZipFileName)));
+
+				Hashtable<String, Integer> indexes = new Hashtable<String, Integer>();
+				byte[] buffer = new byte[1024];
+
+				for (int i = 0; i < mergeThese.size(); i++) {
+					ZipInputStream in = null;
+
+					try {
+						ZipDownloadDetails zdd = mergeThese.get(i);
+						String strTempFile = zdd.getFileName();
+
+						Integer index = indexes.get(zdd.getLayerName());
+						if (index == null) {
+							index = 0;
+						} else {
+							index++;
+							indexes.remove(zdd.getLayerName());
+						}
+
+						indexes.put(zdd.getLayerName(), index);
+
+						String folderName = zdd.getLayerName() + "_" + index + "/";
+						out.putNextEntry(new ZipEntry(folderName));
+
+						in = new ZipInputStream(new FileInputStream(strTempFile));
+						ZipEntry ze = in.getNextEntry();
+						while (ze != null) {
+							String fileName = ze.getName();
+							out.putNextEntry(new ZipEntry(folderName + fileName));
+							int len;
+							while ((len = in.read(buffer)) > 0) {
+								out.write(buffer, 0, len);
+							}
+							ze = in.getNextEntry();
+						}
+
+						out.closeEntry();
+						in.close();
+						deleteFile(strTempFile);
```

Comment:
You can remove in.close() as it's in finally. Also shouldn't the file be deleted in finally to clean up even in failure?